✨ [Feat] CRUD 중 CR 구현

rlawngjs0313 opened the pull request:
## 📍 PR 타입 (하나 이상 선택)
- [x] 기능 추가
- [x] 버그 수정
- [x] 의존성, 환경 변수, 빌드 관련 코드 업데이트
- [ ] 기타 사소한 수정

## ❗️ 관련 이슈 링크
Close #18

## 📌 개요
- Exception Handler 구현
- Article API 구현 (전체 게시물 조회, 게시물 ID를 통해 조회, 게시물 추가)
- Reply API 구현 (게시물 ID를 이용해 댓글 전체 조회, 게시물 ID를 이용해 게시물 댓글 추가)
- 각 API의 응답을 DTO로 응답하도록 구현
- Article 관련 오류 (존재하지 않는 게시물 조회, 존재하지 않는 게시물에 댓글 달기) Exception 추가, ErrorCode 추가
- RequestDTO에는 toEntity()를 구현, ResponseDTO에는 toDTO를 구현

## 🔁 변경 사항
- 149e5e8333a260f7d792534723af691c481652e6
- 94e6cda7a41ee324b120c0696ec96f0d5f8cd031
- b6584ce01a7c8d97d2ea0e5240fc5475215365a2
- 7154936754f5e4db9c0c6a3f45a14ec1e18aff44
- 64ff477f8337b1083fc24f4bbebdca469de3f35c

## 📸 스크린샷
<img width="1163" alt="스크린샷 2025-03-31 01 58 11" src="https://github.com/user-attachments/assets/79a94cf6-49cc-4fb4-82fe-05bc11e8f041" />


## 👀 기타 더 이야기해볼 점

## ✅ 체크 리스트
- [x] PR 템플릿에 맞추어 작성했어요.
- [x] 변경 내용에 대한 테스트를 진행했어요.
- [x] 프로그램이 정상적으로 동작해요.
- [x] PR에 적절한 라벨을 선택했어요.
- [x] 불필요한 코드는 삭제했어요.

rlawngjs0313 committed:
✨ [Feat] 예외 핸들러 생성
🎨 [style] onFailure() 구조 변경

rlawngjs0313 committed:
✨ [Feat] Article 클래스 구현
✨ [Feat] 게시글 생성, 전체 조회, ID로 조회 구현
✨ [Feat] Reply 클래스 구현
✨ [Feat] 댓글 생성, 전체 조회, ID로 조회 구현 (게시글과 매핑 구현 X)
✨ [Feat] 생성 시간, 수정 시간을 담은 BaseEntity 구현

rlawngjs0313 committed:
✨ [Feat] Controller 응답 DTO로 대체
✨ [Feat] DTO -> Entity, Entity -> DTO 구조를 이용해 Service 로직 구현
✨ [Feat] ArticleId를 통해 Reply 조회 기능 구현

rlawngjs0313 committed:
✨ [Feat] Article 관련 ErrorCode, Exception 구현
🐛 [fix] ControllerAdvice와 Swagger 간 충돌 해결
🏗️ [build] Springboot Version 3.4.3 -> 3.3.5 (Swagger 충돌 해결 목적)

rlawngjs0313 committed:
🔨 [refactor] 불필요한 import 삭제

rlawngjs0313 committed:
🎨 [style] ArticleErrorCode 403코드 메세지 변경

Properks commented on src/main/java/com/example/umc8th/article/dto/ArticleResponseDTO.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Builder
+    @Data
+    public static class articleDTO {
```
Java에서 클래스 생성 시 PascalCase 지켜주세요!

Properks commented on src/main/java/com/example/umc8th/article/entity/Article.java:
```diff
@@ -0,0 +1,31 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column
+    private String title;
+    private String content;
+    private int likeNum;
```
다음부터 엔티티를 생성할 때 `@Column`을 이용해서 name을 다 지정해주시면 좋을 것 같아요.
(지정하지 않아서 nativeQuery 사용 시 query와 column 이름이 일치하지 않아서 오류가 났었던 적이 있었습니다.)

Properks commented on src/main/java/com/example/umc8th/global/apiPayload/exception/ArticleException.java:
```diff
@@ -0,0 +1,14 @@
+import com.example.umc8th.global.apiPayload.code.BaseErrorCode;
+import lombok.Getter;
+
+@Getter
+public class ArticleException extends RuntimeException {
+
+    private final BaseErrorCode code;
+
+    public ArticleException(BaseErrorCode code) {
+        this.code = code;
+    }
+}
```
ArticleException은 global이 아닌 article 패키지 안에 exception 패키지를 하나 더 만들어서 관리하는 것이 추후에 Exception 클래스가 많아질 때 관리하기 수월합니다!

Properks commented on src/main/java/com/example/umc8th/global/apiPayload/exception/handler/ExceptionAdvice.java:
```diff
@@ -0,0 +1,31 @@
+        GlobalResponse<String> response = GlobalResponse.onFailure(code.getCode(), code.getMessage());
+        return ResponseEntity.status(code.getStatus()).body(response);
+    }
+    @ExceptionHandler(ArticleException.class)
+    public ResponseEntity<?> articleException(ArticleException e) {
+        log.warn("Article exception: {}", e.getCode().getMessage());
+        BaseErrorCode code = e.getCode();
+        GlobalResponse<String> response = GlobalResponse.onFailure(code.getCode(), code.getMessage());
+        return ResponseEntity.status(code.getStatus()).body(response);
+    }
```
위의 코드와 중복되기도 하고 위의 코드리뷰로 코드를 변경하게 된다면 ArticleException 대신 모든 에러의 최상위 클래스인 Exception에 대한 Handler를 만들어주세요!

Properks reviewed (approved):
전체적으로 구조가 다를 뿐 코드 상의 문제는 없습니다. 다만 연습 과정이기에 되도록 수정 후 머지하는 것을 추천드립니다!

rlawngjs0313 committed:
🔨 [refactor] ArticleDTO 파스칼 케이스에 맞게 수정

rlawngjs0313 committed:
🔨 [refactor] import 최적화
🔨 [refactor] ArticleException Article 패키지로 리팩토링
✨ [feat] Exception Handler 추가
🔨 [refactor] @Column name 추가

rlawngjs0313 commented:
수정 완료했습니다.
- https://github.com/SMUMC-8th/SPRING_STUDY/pull/21/commits/f5f79d05402a57b1ddfc355ad7ff2e13d722364e
- https://github.com/SMUMC-8th/SPRING_STUDY/pull/21/commits/e4e03a040fbd3457d73508773adc9e76c2f13dc3

Properks commented on src/main/java/com/example/umc8th/article/exception/handler/ArticleExceptionAdvice.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@Slf4j
+@RestControllerAdvice(annotations = RestController.class)
+public class ArticleExceptionAdvice {
+
+    @ExceptionHandler(ArticleException.class)
+    public ResponseEntity<GlobalResponse<String>> articleException(ArticleException e) {
+        log.warn("Article exception: {}", e.getCode().getMessage());
+        BaseErrorCode code = e.getCode();
+        GlobalResponse<String> response = GlobalResponse.onFailure(code.getCode(), code.getMessage());
+        return ResponseEntity.status(code.getStatus()).body(response);
+    }
+}
```
이 부분은 ArticleException에 대한 처리인데 원래는 ArticleException을 GeneralException을 상속해서 진행하는데 만약 이렇게 ArticleExceptionAdvice를 만들게 되면 도메인 별로 만들어야되서 많이 복잡해질 수 있을 것 같습니다!

또한 내용이 비슷하기도 한데 만약 Exception이 터진 정확한 위치를 알고 싶으신거면 ExceptionAdvice의 GeneralException을 받는 메소드에서 log 찍는 코드를 아래처럼 바꿀 수도 있습니다.
```java
log.error("{}: {}", e.getClass().getSimpleName(), message);
```

Properks commented on src/main/java/com/example/umc8th/article/exception/ArticleException.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Getter;
+
+@Getter
+public class ArticleException extends RuntimeException {
+
+    private final BaseErrorCode code;
+
+    public ArticleException(BaseErrorCode code) {
+        this.code = code;
+    }
+}
```
```java
@Getter
public class ArticleException extends GeneralException {
    public ArticleException(ArticleErrorCode code) {
        super(code)
    }
}
```
이렇게 만들면 하나의 ExceptionAdvice에서 다 받을 수 있고 도메인마다 추가하지 않아도 되어서 좋습니다!

rlawngjs0313 committed:
🔨 [refactor] ArticleException 구조 변경, Advice 구조 변경

rlawngjs0313 commented:
- b8112c60499d65ee2ad9f647776caa018c006115
- 구조를 이렇게 변경하면 나중에 API를 추가할 경우에 해당 Exception을 상위 Exception을 상속하고 ErrorCode만 작성하면 되는 걸로 이해하면 될까요?
- 나중에 따로 로직을 거쳐 처리해야 하는 경우에만 @ExceptionHandler로 추가하면 될까요?
